sofkasim.py

```diff
@@ -50,58 +60,65 @@ def perform(self):
                 n = int(c[:2], 8)
                 v = int(c[2:], 8)
                 device = get_device(int(c[:2], 8))
-                dprint(device, c)
+                dprint(f'\n{device} {c}')
                 if n == 62:  # Interrupt timer
                     self.clock = v
                 if 0 < n < 4:  # Osc
                     dprint('OSC', n)
                     if self.oscillators[n - 1]:
                         self.oscillators[n - 1].change(v)
-                    else:
-                        self.oscillators[n - 1] = Oscillator(v)
-                    self.active = n - 1
                 if 23 < n < 27:  # Envelopes
                     n = n - 24
                     dprint('Envelope', n + 1)
                     t = self.current_time
-                    d = self.secs(v)
                     if self.envelopes[n]:
-                        self.envelopes[n].addstage(t, d)
+                        self.envelopes[n].addstage(t, v)
                     else:
-                        self.envelopes[n] = Envelope(t, d)
-                    self.current_time += d
-                    self.oscillators[self.active].addtime(d)
+                        self.envelopes[n] = Envelope(t, v)
                 if n == 60:  # Wait timer
                     d = self.secs(v)
                     dprint('WAIT:', d)
                     self.current_time += d
-                    self.oscillators[self.active].addtime(d)
+                    [osc.addtime(d) for osc in self.oscillators if osc]
+
         # Write the generated audio
-        sources = [o for o in self.oscillators if o]
-        sources += [e for e in self.envelopes if e]
-        dprint('SOURCES', sources)
-        output = ' '.join(['(seq %s)' % ' '.join(s.out()) for s in sources])
-        return f'(mult {output})'
+        # TODO:
+        # mix oscillators, envelopes, and their amplifiers as patched
+        # mix signals
+        signals = []
+        for i, osc in enumerate(self.oscillators):
+            if osc.pitch == 0 and not osc.history:
+                continue
+            if env := self.envelopes[i]:
+                signals.append(f'(mult {osc.out()} {env.out()})')
+            else:
+                signals.append(osc.out())
+        dprint('SIGNALS', signals)
+        #output = ' '.join(['(seq %s)' % ' '.join(s.out()) for s in sources])
+        output = ' '.join(signals)
+        return f'(sim {output})'
 
-    def secs(self, n):
+    def secs(self, n) -> float:
         """ Number of seconds of time with current clock."""
         return n * 1 / self.clock
 
 
 class Envelope:
-    def __init__(self, current_time, duration):
-        self.stages = []  # list of (time, duration) for alternating attack / decay
+    def __init__(self, current_time, duration:int):
+        self.stages = []  # list of (time, duration in seconds) for alternating attack / decay
```

Comment:
this looks unclear -- `duration:int` followed by  `# list of (time, duration in seconds)`  
The first duration is a 11 bit integer, representing position in a  control range, the second is that int value converted into seconds (a float). Not a code error, just needs clarification in the docs.